### PR TITLE
ARROW-16527: [Gandiva][C++] Add binary functions

### PR DESCRIPTION
**Pull request by Johnnathanalmeida**
This PR was created to implement binary functions in Gandiva side based on [Hive implementation](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFToBinary.java).

This PR implements the follow signatures:

FunctionSignature{name =binary, return type =binary, param types =[string]}
FunctionSignature{name =binary, return type =binary, param types =[binary]}

### CODE REVIEW

**Comment by Johnnathanalmeida** [comment]
@github-actions crossbow submit java-jars

File: cpp/src/gandiva/precompiled/string_ops.cc [hunk 1]
```diff
@@ -705,6 +705,25 @@ CAST_VARCHAR_FROM_VARLEN_TYPE(binary)
 CAST_VARBINARY_FROM_STRING_AND_BINARY(utf8)
 CAST_VARBINARY_FROM_STRING_AND_BINARY(binary)
 
+#define CAST_BINARY_FROM_STRING_AND_BINARY(TYPE)                                         \
+  GANDIVA_EXPORT                                                                         \
+  const char* castBINARY_##TYPE(gdv_int64 context, const char* data, gdv_int32 data_len, \
+                                int32_t* out_length) {                                   \
+    int32_t len = static_cast<int32_t>(data_len);                                        \
+    if (len < 0) {                                                                       \
```
Comment (projjal) [inline cpp/src/gandiva/precompiled/string_ops.cc:713]: this check is not required

Comment (Johnnathanalmeida) [reply cpp/src/gandiva/precompiled/string_ops.cc]: Ok, this if clause was removed.

File: cpp/src/gandiva/function_registry_string.cc [hunk 2]
```diff
@@ -439,6 +439,14 @@ std::vector<NativeFunction> GetStringFunctionRegistry() {
                      kResultNullIfNull, "right_utf8_int32",
                      NativeFunction::kNeedsContext),
 
+      NativeFunction("binary", {}, DataTypeVector{binary()}, binary(), kResultNullIfNull,
+                     "castBINARY_binary",
+                     NativeFunction::kNeedsContext | NativeFunction::kCanReturnErrors),
```
Comment (projjal) [inline cpp/src/gandiva/function_registry_string.cc:444]: It doesn't requried context and canreturnerrors flags.

Comment (Johnnathanalmeida) [reply cpp/src/gandiva/function_registry_string.cc]: Ok, these two flags was removed.

**Review by projjal — APPROVED** [review]

**Comment by Johnnathanalmeida** [comment]
@github-actions crossbow submit java-jars

**Pull request closed without merging**